### ISSUE TITLE
Regression in 0.0.82: following a recursive list in a loop introduces Unknown

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#4555]
### Summary

This example passes in ty 0.0.81 and fails in 0.0.82:

```python
Capybara = list['Capybara'] | int


def hutia(gundi: Capybara) -> Capybara:
    while isinstance(gundi, list):
        gundi = gundi[0]
    return gundi
```

Run either released binary with:

```console
ty check --error unsound-return-statement example.py
```

Expected: `gundi` remains a `Capybara`, and after the loop can only be an `int`. Indexing the recursive list does not introduce an untyped value.

Actual in 0.0.82: `unsound-return-statement` at `return gundi`, with `Unknown` in the inferred type. 0.0.81 reports no diagnostics.

The regression disappears if the loop becomes an `if`, or if the alias becomes nonrecursive (`list[int] | int`). Annotating the loop assignment (`gundi: Capybara = gundi[0]`) also makes ty accept the example. In a function that first assigns the loop variable and subsequently annotates its recursive assignment, mypy rejects the workaround as a redeclaration; annotating only the initial assignment does not avoid the ty regression.

Minimized systematically from a JSON traversal function, checking both releases after each reduction. No third-party dependencies or typing imports remain.

Potentially related: [#4221](https://github.com/astral-sh/ty/issues/4221), involving a different cyclic inference path, and the closed [#2872](https://github.com/astral-sh/ty/issues/2872). I did not find an exact duplicate; these may share an underlying cause.
